Document contributing (#2675)
agrc/gis.utah.gov: .github/collaborating.md

Staff may **close** any issue or pull request they believe is not relevant to the future of the UGRC project.
Where this is unclear, the issue should be left open for several days for additional discussion.
Where this does not yield input from UGRC Staff or additional evidence that the issue has relevance, then the issue may be closed.
Remember that issues can always be re-opened if necessary.

> \[!IMPORTANT]\
> We recommend staff to avoid Updating/Rebasing PRs unnecessarily, since we use [GitHub Merge Queues](https://github.blog/2023-07-12-github-merge-queue-is-generally-available/)
> to merge Pull Requests, which automatically rebases and runs CI-checks against the latest base branch.

## Accepting modifications

All UGRC code and documentation modifications should be performed via GitHub pull requests.
Only the Website Team can merge their work and should do so carefully.

All pull requests must be reviewed and accepted by another staff member with sufficient expertise who can take full responsibility for the change.

Pull Requests can only be merged after all CI Checks have passed.

> In some cases, it may be necessary to summon a qualified staff member to a pull request for review by @-mention.

Reviewer: I love the use of "summon"! :)

If you are unsure about the modification and are not prepared to take full responsibility for the change, defer to another staff member.

We recommend staff follow the guidelines on the [Contributing Guide](./CONTRIBUTING.md#before-merging) for reviewing and merging Pull Requests.

## Technologies used in the website

> The UGRC Website is built upon [React][] and [Astro][], the UI Rendering Engine and the Framework that builds the Website;

Reviewer: ```suggestion
The UGRC Website is built upon [Astro][], the Framework that builds the Website, and [React][] the UI Rendering Engine.
```

The Website also uses several other Open Source libraries (not limited to) listed below:

> - Styling is done with [tailwindcss][].
>   - Tailwind is used as our CSS Framework and the Foundation of our Design System
> - [Hero Icons](https://heroicons.com/) is an SVG Icon Library used within our Codebase
> - [Radix UI][] is a collection of customizable UI components
> - [Shiki][] is a Syntax Highlighter used for our Code boxes
>   - The syntax highlighting is done within the processing of the Markdown files with the MDX compiler as a Rehype plugin.
> - [MDX][] and Markdown are used for structuring the Content of the Website
> - We use [Rehype](https://github.com/rehypejs/rehype) and [Remark](https://github.com/remarkjs/remark) to extend MDX functionality

Reviewer: ```suggestion
- Styling is done with [tailwindcss][].
  - Tailwind is used as our CSS Framework and the Foundation of our Design System.
- [Hero Icons](https://heroicons.com/) is an SVG Icon Library used within our Codebase.
- [Radix UI][] is a collection of customizable and accessible UI components.
- [Shiki][] is a Syntax Highlighter used for our Code boxes.
  - The syntax highlighting is done within the processing of the Markdown files with the MDX compiler as a Rehype plugin.
- [MDX][] and Markdown are used for structuring the content of the website.
- We use [Rehype](https://github.com/rehypejs/rehype) and [Remark](https://github.com/remarkjs/remark) to extend MDX functionality.
- Our styling for the website is based on the [Utah Design System](https://designsystem.utah.gov/).
```

## Code Editing

### Structure of this repository

- Astro components are defined on `/src/components/`
- React components are defined on `/src/components/react`
- Astro layouts are defined on `/src/layouts`
- The global stylesheet is declared in `/src/styles/tailwind.css`
- Public files are stored on `/public`
- Files for download are store in `/src/downloads`
  - The files are then organized by pillar
- Website content is defined within `/src/pages`
  - All content is in the Astro format.
- Website blog content is defined within `/src/content/blog`
  - All content is in Markdown or MDX.
  - The top of each Markdown file is a YAML (Frontmatter) block for page-specific information passed to various templates.
- Typescript import aliases are defined in `/tsconfig.json`

> - Multi-Purpose javascript functions are defined on `/src/utilities`

Reviewer: ```suggestion
- Multi-Purpose JavaScript functions are defined on `/src/utilities`
```

```astro
---
import Layout from '@layouts/FullWidth.astro';

import type { IStandardPageMetadata } from '@models/types';

import QuickLinks from '@components/page/QuickLinks.astro';
import Section from '@components/page/Section.astro';

import { Image } from 'astro:assets';

import myPhoto from '@images/about/me.svg';

const page: IStandardPageMetadata = {
  pageTitle: 'Title of the page',
  pageDescription: '110-160 characters describing the page for SEO',
  section: [{ title: 'An optional section title' }],
};
---

<Layout pageTitle={page.pageTitle}>
  <QuickLinks links={page.section} slot="quick-links" />
  <Section {...page.section[0]}>
    <Image
      slot="image"
      src={myPhoto}
      loading="eager"
      alt="A gender neutral description that will be read to non sighted people"
      class="w-full max-w-[400px] rounded-2xl"
    />
    <p>Page content</p>
  </Section>
</Layout>
```

### Translating Pages

See the [Translation Guidelines](./TRANSLATION.md) for the website translation policy.

## Creating Astro Components

The UGRC Website uses [Astro][] as a Frontend Library to develop the Website.

> Astro allows us to create user interfaces with a modern take on Web Development falling back to [React][] if necessary using dynamic islands.
> 
> If you're unfamiliar with React or Web Development in general, we encourage a read before taking on complex issues and tasks as this repository is **not for educational purposes** and we expect you to have a basic understanding of the technologies used.
> 
> We also recommend getting familiar with technologies such as [Astro][], [MDX][], and [tailwindcss][].

Reviewer: ```suggestion
Astro allows us to create user interfaces with a modern take on Web Development falling back to [React][] if necessary using dynamic islands.

If you're unfamiliar with Astro, React, or Web Development in general, we encourage you to learn these technologies on your own before taking on complex issues and tasks. This repository is **not for educational purposes** and we expect you to have a basic understanding of the technologies used.

We also recommend getting familiar with technologies such as [Astro][], [MDX][], and [tailwindcss][].
```

```astro
---
type Props {
  title: string;
}

const { title } = Astro.props;
---

<div>
    <slot />
</div>

```

### Best practices for Component development in general

- Only spread props `{ ... }` on the definition of the Component (Avoid having a variable named `props`)
- When importing types, use `import type { NameOfImport } from 'module'`

## Remarks on Technologies Used

The UGRC Website is a somewhat complex application and at times non-trivial solutions have been implemented to solve certain technical challenges.
Historical decision making can be largely found on past issues, conversations on Slack and GitHub discussions. However, we also wish to highlight some of the notable development decisions that we have made here.

### Why Astro?

We've found that Astro is simply versatile, hackable, stable, community-maintained and has a great ecosystem.
The reasoning goes deeper, but as a long-term Framework it is the most suitable choice.

#### What is `/public/_redirects`?

This file is responsible for defining the rewrite rules for the Website when deployed to Netlify.
It is used for defining redirects. (Such as Internal Redirects and External ones).

> Redirect operations change the original URL to a new one.

Reviewer: ```suggestion
Redirect operations change the original URL to a new one.

Please run [/src/scripts/format-redirect.mjs](./src/scripts/format-redirect.mjs) after making changes to this file to preserve the formatting.
```
